File: python/134_Gas_Station.py
import unittest

class Solution:
    def solve(self, gas, cost):
        return self.canCompleteCircuit(gas, cost)
        
    # @param {integer[]} gas
    # @param {integer[]} cost
    # @return {integer}
    def canCompleteCircuit(self, gas, cost):
        length=len(gas)
        index,end=length-1,0
        contain=0
        while index>=end:
            if contain+gas[end]-cost[end]>=0:
                contain=contain+gas[end]-cost[end]
                end=end+1
            else:
                contain=contain+gas[index]-cost[index]
                index=index-1
        return (index+1)%length if contain>=0 else -1
        
        
class Test(unittest.TestCase):
    #override
    def setUp(self):
        self.s=Solution()
        
    def tearDown(self):
        print "TestComplete"
        
    def genTestCase(self):
        case=[
        ([4],[5]),#-1
        ([1,2,2],[1,2,2]),#0
        ([1,2,2],[1,2,3]),#-1
        ([1,3,2],[1,1,3]),#0
        ([8,2,4],[7,4,1]),#-1
        ]
        self.cases=case
        
    def test(self):
        self.genTestCase()
        for case in self.cases:
            print "testcase: ", case[0],case[1]
            res=self.s.solve(case[0],case[1])
            print "result: ",res
            # if assert is needed
            
unittest.main()

File: Java/134_Gas_Station.java
public class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        if(gas==null||gas.length<1)return -1;
        int len=gas.length;
        int start=len-1,end=0, curgas=0;// 从第len-1个站点开始，start指向第len-1个站点，end指向下一个要到达的站点
        // 如果存在一个环可以走完所有站点，那么在这个环内，从任意一点开始，累积到起点的所得gas，一定可以覆盖该点以后的所有站点，因此可以在开始时任意选择一个点，当gas不足时往前寻找gas补充。这种方法能保证在有解的情况下找到一个解。
        while(start>=end){// 每一步不是起点前移就是终点后移，由于起终点的初始值选择，这里一旦两者相遇则说明已经访问了一圈
        //对于每一步，只需判断是向前走还是向后退。
            if(curgas+gas[end]-cost[end]>=0){// 若当前存留的汽油加上当前站点的汽油足以到达下一个站点
                curgas+=gas[end]-cost[end];// 则到达下一个站点
                end++;
            }else{// 否则向起点前一个站点索求汽油
                curgas+=gas[start]-cost[start];
                start--;
            }
        }
        return curgas>=0?(start+1)%len:-1;
    }
}